lib/swagcov/coverage.rb

```diff
@@ -13,20 +15,25 @@ def initialize
     end
 
     def report
-      Rails.application.routes.routes.each do |route|
-        # https://github.com/rails/rails/blob/48f3c3e201b57a4832314b2c957a3b303e89bfea/actionpack/lib/action_dispatch/routing/inspector.rb#L105-L107
-        # Skips route paths like ["/rails/info/properties", "/rails/info", "/rails/mailers"]
-        next if route.internal
+      collect_coverage
+      routes_output(@routes_covered, "green")
+      routes_output(@routes_ignored, "yellow")
+      routes_output(@routes_not_covered, "red")
 
-        # Skips routes like "/sidekiq"
-        next unless route.verb.present?
+      final_output
+
+      exit @total - @covered
+    end
+
+    private
+
+    attr_reader :dotfile
 
+    def collect_coverage
```

Comment:
I think that the main debate around testing private method is not exactly our kind of debate. When trying to test this what you encounter is more of a design problem. I think there will be less of dilemmas if we were to split this class in two. Or even three. Because this class is doing at least two things. You tried making one of those things private but it's the one that requires more testing.

If I were to really force that «S» in S.O.L.I.D., I'd suggest to have `AppRoutesProcessor`, `CoverageCollector` and `CoverageReporter`. (Maybe more human and less Java™ Enterprise names would be better.) In that case most of the edge cases could be tested more thoroughly in isolation. And you only need a few integration tests.

If you liked what extracting `Dotfile` did to the application, I think you might like pulling those parts out too.

Reply:
I agree 💯 . I definitely realized this when writing tests 🙃